compiler/driver.ml

```diff
@@ -295,10 +313,16 @@ let driver source_file (options : Cli.options) : int =
               results
           | (`OCaml | `Python | `Lcalc | `Scalc | `Plugin _) as backend -> (
             Cli.debug_print "Compiling program into lambda calculus...";
+            Cli.debug_format "before:";
+            Format.print_cut ();
+            Cli.debug_format "%a" (Shared_ast.Program.format ~debug:true) prgm;
             let prgm =
               if options.avoid_exceptions then
-                Lcalc.Compile_without_exceptions.translate_program prgm
-              else Lcalc.Compile_with_exceptions.translate_program prgm
+                Shared_ast.Program.untype
+                @@ Lcalc.Compile_without_exceptions.translate_program prgm
+              else
+                Shared_ast.Program.untype
+                @@ Lcalc.Compile_with_exceptions.translate_program prgm
```

Comment:
for with exception, typing must be done after optimization in some cases (when some variables are left undefined). This is because empty is polymorphic hence of type any if not used.